packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js

```diff
@@ -10,6 +32,146 @@ class MultiFileUploadQuestion extends Question {
+					journeyResponse.answers[this.fieldName].uploadedFiles,
+					removedFiles
+				);
+			}
+
+			// save files to blob storage
+			const uploadedFiles = await this.#saveFilesToBlobStorage(req, journeyResponse);
+
+			// add saved docs to response
+			responseToSave.answers[this.fieldName].uploadedFiles = [];
+
+			if (journeyResponse?.answers[this.fieldName]?.uploadedFiles) {
+				responseToSave.answers[this.fieldName].uploadedFiles.push(
+					...journeyResponse.answers[this.fieldName].uploadedFiles
+				);
+			}
+
+			if (uploadedFiles) {
+				responseToSave.answers[this.fieldName].uploadedFiles.push(...uploadedFiles);
+			}
+
+			// save answer to database
+			await patchQuestionResponse(journeyResponse.journeyId, encodedReferenceId, responseToSave);
+
+			// todo: validation after saving required?
+			// if (Object.keys(errors).length > 0) {
+			// 	const answer = journeyResponse.answers[this.fieldName].uploadedFiles;
+			// 	let viewModel = journey.getQuestionViewModel(sectionObj, this, answer);
+			// 	viewModel.errors = errors;
+			// 	// multi-file upload validation would otherwise map these errors individual to e.g.
+			// 	// `#files.supporting-documents[3]` which does not meet the gov uk presentation requirements.
+			// 	viewModel.errorSummary = errorSummary.map((error) => ({
+			// 		...error,
+			// 		href: '#'
+			// 	}));
+			// 	return this.#showError(res, viewModel);
+			// }
+
+			// map the response back to journeyResponse
+			journeyResponse.answers[this.fieldName] = responseToSave.answers[this.fieldName];
+
+			// update journey based on response
+			const updatedJourney = new journey.constructor(journeyResponse);
+
+			// this is the `name` of the 'upload' button in the template.
+			// when is this used?
+			if (body['upload-and-remain-on-page']) {
+				return res.redirect(
+					updatedJourney.getCurrentQuestionUrl(sectionObj.segment, this.fieldName)
+				);
+			}
+
+			return res.redirect(
+				updatedJourney.getNextQuestionUrl(sectionObj.segment, this.fieldName, false)
+			);
+		} catch (err) {
+			logger.error(err);
+
+			return this.#showError(res, journey, sectionObj, {
+				errorSummary: [{ text: 'Failed to upload files', href: '#' }]
+			});
+		}
+	};
+
+	async #saveFilesToBlobStorage(req, journeyResponse) {
+		let result = [];
+
+		if (this.fieldName in req.files) {
+			// const validFiles = getValidFiles(errors, body.files[this.fieldName]);
+			// eslint-disable-next-line no-restricted-syntax
+			for await (const file of req.files[this.fieldName]) {
+				const document = await createDocument(
+					{
+						id: journeyResponse.journeyId,
+						referenceNumber: journeyResponse.referenceId
+					},
+					file,
+					file.name,
+					this.documentType.name
+				);
+
+				const mappedDoc = mapMultiFileDocumentToSavedDocument(document, document?.name, file.name);
+
+				result.push(mappedDoc);
+			}
+		}
+
+		return result;
+	}
+
+	#showError(res, journey, sectionObj, customViewData) {
```

Comment:
Would this work at a level above? As in should we implement showError on the base question? Perhaps we should be doing this in every scenario where we save data (e.g when cosmos fails to save an answer)?

Reply:
Ah I can see you're not handling this with the render page method and passing the error summary in there in the catch block, looks good 👍